articles/graph-traversals-java/index.md

```diff
@@ -0,0 +1,204 @@
+---
+
+title: Graph Implementation and Traversal Algorithms (Java)
+
+---
+Graphs are one of the most common data structures in computer science. Graphs are made up of nodes and edges. There are many applications for graph structures in the real world including relationships (Facebook), locations (Google Maps), programming analysis, and more.
+
+<!-- <img src="graph.png" alt="drawing" width="600"/> -->
+
+![graph](graph.jpg)
+
+By the end of this article, readers should know
+1. What graphs are
+2. How graphs are implemented in code
+3. How to traverse a graph
+    * Breadth first search iterative
+    * Depth first search iterative and recursive
+
+<br />
+
+## Types of graphs
+
+There are different types of graphs like undirected, directed, unweighted, unweighted, etc. All graphs have nodes and edges, but there are different structures and properties between different types. Undirected graphs have directionless edges between nodes while directed graphs have edges that point in a direction. Directed graphs can only be traversed in the direction the edges point. Unweighted graphs have zero value edges while weighted graphs have non zero value edges. These values can be positive or negative.
+
+<br />
+
+Undirected and Unweighted  |  Directed and Weighted	 |	
+:-------------------------:|:-----------------------:|
+|![undirected](undirected.png)|![directed](directed.png)
+
+<br />
+<br />
+
+## How to implement a graph
+
+### Node object
+
+First, we create a Node object. It contains a value and an array list of neighboring nodes.
+
+```
+import java.util.ArrayList;
+import java.util.List;
+
+public class Node {
+	
+	int value;
+	List<Node> neighbors;
+
+	public Node(int value) {
+		this.value = value;
+		neighbors = new ArrayList<>();
+	}
+	
+	public void addEdge(Node to) {
+		neighbors.add(to);
+	}
+}
+
+```
+### Graph object
+
+Next, we implement a graph object. The code below will create an undirected graph as shown before.
+
+```
+public class Graph {
+    public static void main(String[] args) {
+        Node n0 = new Node(0);
+        Node n1 = new Node(1);
+        Node n2 = new Node(2);
+        Node n3 = new Node(3);
+        Node n4 = new Node(4);
+        Node n5 = new Node(5);
+        Node n6 = new Node(6);
+	
+        n0.addEdge(n1);
+        n1.addEdge(n0);
+        n1.addEdge(n3);
+        n1.addEdge(n2);
+        n2.addEdge(n1);
+        n2.addEdge(n4);
+        n3.addEdge(n1);
+        n3.addEdge(n4);
+        n3.addEdge(n5);
+        n4.addEdge(n2);
+        n4.addEdge(n3);
+        n5.addEdge(n3);
+
+        // Uncomment any method and run
+        // bfs(n0);
```

Comment:
Create a new section at the end and call it outputs. Under that section, copy the outputs from the terminal and paste it as a code formatted snippet. Use ```bash to format the output as seen on the terminal. For each of the methods discussed here.